### PR TITLE
support basic authentication. Resolves issue #493

### PR DESCRIPTION
**Pull request by TonyLHansen**

### CODE REVIEW

**Comment by TonyLHansen** [comment]
see the description in issue #493

**Comment by vpetersson** [comment]
@TonyLHansen looks like the tests are failing. :(

**Comment by TonyLHansen** [comment]
Ok, pep8 and compilation checks are all passing now.

I hope this is acceptable.

**Comment by vpetersson** [comment]
@over64 / @femdom - Can you guys take a look?

**Comment by over64** [comment]
Does it means that users will hit on login form after upgrade? They can be affraid so much after this :fire: Because they has no assumptions about credentials. Can we make this feature optional?

**Comment by vpetersson** [comment]
Yes, this must be an opt-in feature. Not a mandatory one.

**Comment by TonyLHansen** [comment]
The use of credentials is **totally optional**. If you put in a username and password into the configuration file, they will be prompted for. If you don't, they will not be prompted for.

**Comment by TonyLHansen** [comment]
I'm not sure what the status of this issue is. It's marked closed, but it doesn't look like it was merged into the master branch.

**Comment by vpetersson** [comment]
@TonyLHansen This ticket is not closed. The issue you initially referred to it in is. The issue is still pending review before it can be merged.

**Comment by TonyLHansen** [comment]
Sorry, my apologies. PEBKAC

File: settings.py [hunk 1]
```diff
@@ -115,4 +119,20 @@ def get_listen_ip(self):
     def get_listen_port(self):
         return self['listen'].split(':')[1]
 
+    def check_user(self, user, pswd):
+        try:
+            au = self['user']
+            pw = self['password']
+            logging.info("check_user(" + user + "," + pswd + "), looking for (" + au + "," + pw + ")")
```
Comment (TonyLHansen) [inline settings.py]: Agreed -- that was mostly for debugging. Do you want me to remove the ,pswd, or just remove them when it's integrated? An alternative is to log the md5 hash of the passwords, which IS an acceptable practice.

File: settings.py [hunk 2]
```diff
@@ -115,4 +119,20 @@ def get_listen_ip(self):
...
+            if au == "" and pw == "":
+                logging.info("check_user() requires no password")
+                return True
+            ret = au == user and pw == pswd
+            logging.info("check_user() is returning %s" % ret)
+            return ret
+        except:
+            e = sys.exc_info()[0]
+            logging.info("check_user could not find a user " + str(e))
```
Comment (TonyLHansen) [inline settings.py]: agreed

File: server.py [hunk 3]
```diff
@@ -102,6 +102,9 @@ def template(template_name, **context):
 # Model
 ################################
 
+def check(user, passwd):
+    return settings.check_user(user, passwd)
```
Comment (TonyLHansen) [inline server.py]: trying to preserve separation of functionality.  It's a stylistic decision.

File: settings.py [hunk 4]
```diff
@@ -115,4 +119,20 @@ def get_listen_ip(self):
...
+    def check_user(self, user, pswd):
+        try:
+            au = self['user']
+            pw = self['password']
+            logging.info("check_user(" + user + "," + pswd + "), looking for (" + au + "," + pw + ")")
+            if au == "" and pw == "":
+                logging.info("check_user() requires no password")
+                return True
+            ret = au == user and pw == pswd
+            logging.info("check_user() is returning %s" % ret)
```
Comment (TonyLHansen) [inline settings.py]: agreed. I'll change to logging.debug()

Comment (TonyLHansen) [inline settings.py]: and I reduced the number of logging calls

File: settings.py [hunk 5]
```diff
@@ -115,4 +119,20 @@ def get_listen_ip(self):
     def get_listen_port(self):
         return self['listen'].split(':')[1]
 
+    def check_user(self, user, pswd):
+        try:
+            au = self['user']
+            pw = self['password']
+            logging.info("check_user(" + user + "," + pswd + "), looking for (" + au + "," + pw + ")")
+            if au == "" and pw == "":
```
Comment (TonyLHansen) [inline settings.py]: No, "not au and not pw" gives a different answer if either were a value that evaluated as false. Of course, if everything is working properly, au and pw are strings and your comment is correct. I was being more conservative.

**Comment by TonyLHansen** [comment]
I'll post an updated PR soon.

**Comment by vpetersson** [comment]
@TonyLHansen there's already an updated PR here #533 as we thought this one was abandoned.

**Comment by TonyLHansen** [comment]
sorry, end of year at the day job occurred

**Comment by vpetersson** [comment]
No worries. Please take a look at the other PR and comment there if you find something.

**Comment by TonyLHansen** [comment]
I just reviewed #533 and am fine with the changes.

**Comment by vpetersson** [comment]
👍

**Comment by emanuelstanciu** [comment]
@TonyLHansen I've tried to follow these thread and understand the screenly feature developed here.
Is this just for protecting the admin page with user/pass, or accessing a webpage with user/pass?

My use case is to access kibana which is served on HTTPS with basic auth. I can normally access it https://user:pass@url, but that fails and I've seen various messages about that.

How can I display that with Screenly? Thank you very much for your help.

**Comment by vpetersson** [comment]
@emanuelstanciu That's different. This is for protecting the actual installation. You're referring to accessing a protected site. What you are requesting is already in the development branch and will make it into the next stable release.

**Comment by emanuelstanciu** [comment]
@vpetersson thanks for your quick reply. great stuff.. I'm willing to install the dev version once I get asked.
How do I achieve this then? https://user:pass:url or some other way?

Thanks again for your help.

**Comment by vpetersson** [comment]
@emanuelstanciu Yup. For more info, see #515.

**Comment by C-Duv** [comment]
Could not find much documentation about how to use this feature (apart from the "[Screenly OSE 0.10 is out](https://news.screenly.io/screenly-ose-0-10-is-out-a34978d98a54)" blog post, but in my case I had no `[auth]` section in my configuration file), so for anyone interested, here is how to enable it:
```Shell
cat >> ~/.screenly/screenly.conf <<'EOT'

[auth]
user = foo
password = bar
EOT
```

**Comment by vpetersson** [comment]
Good catch, @C-Duv. We should add some documentation for this.

**Comment by C-Duv** [comment]
Here is how to set authentication when `user` and `password` entries already exists in `~/.screenly/screenly.conf` (which is the case on new installations since v0.10):
```Shell
sed --in-place \
    -e 's/^user\s*=\s*.*/user = foo/' \
    -e 's/^password\s*=\s*.*/password = bar/' \
    ~/.screenly/screenly.conf
```

**Comment by C-Duv** [comment]
@vpetersson Shall I file a PR adding such information to the README.md (under a "Configuration" part)?

**Comment by vpetersson** [comment]
@C-Duv Let's do a PR against a new document in the `docs` folder.

Thanks.

**Pull request merged**